server/auth/types/proxy/proxy_auth.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { get } from 'lodash';
+import { SecurityPluginConfigType } from '../../..';
+import {
+  SessionStorageFactory,
+  IRouter,
+  IClusterClient,
+  CoreSetup,
+  AuthenticationHandler,
+} from '../../../../../../src/core/server';
+import { SecuritySessionCookie } from '../../../session/security_cookie';
+import { SecurityClient } from '../../../backend/opendistro_security_client';
+import { User } from '../../user';
+import { ProxyAuthRoutes } from './routes';
+
+export class ProxyAuthentication {
+  private readonly authType: string = 'proxycache';
+  private readonly securityClient: SecurityClient;
+
+  private readonly userHeaderName: string;
+  private readonly roleHeaderName: string;
+
+  constructor(
+    private readonly config: SecurityPluginConfigType,
+    private readonly sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    private readonly router: IRouter,
+    private readonly esClient: IClusterClient,
+    private readonly coreSetup: CoreSetup
+  ) {
+    this.securityClient = new SecurityClient(this.esClient);
+
+    this.userHeaderName = this.config.proxycache?.user_header?.toLowerCase() || '';
+    this.roleHeaderName = this.config.proxycache?.roles_header?.toLowerCase() || '';
+
+    this.setupRoutes();
+  }
+
+  authHandler: AuthenticationHandler = async (request, response, toolkit) => {
+    let cookie = await this.sessionStorageFactory.asScoped(request).get();
+    const authHeaders: any = {};
+
+    const customProxyHeader = this.config.proxycache?.proxy_header;
+    if (
+      customProxyHeader &&
+      !request.headers[customProxyHeader] &&
+      this.config.proxycache?.proxy_header_ip
+    ) {
+      // TODO: check how to get remoteIp from KibanaRequest and add remoteIp to this header
+      authHeaders[customProxyHeader] = this.config.proxycache!.proxy_header_ip;
+    }
+
+    if (cookie) {
+      if (get(cookie.credentials, this.userHeaderName)) {
+        authHeaders[this.userHeaderName] = cookie.credentials[this.userHeaderName];
+        if (get(cookie.credentials, this.roleHeaderName)) {
+          authHeaders[this.roleHeaderName] = cookie.credentials[this.roleHeaderName];
+        }
+        if (get(cookie.credentials, 'x-forwarded-for')) {
+          authHeaders['x-forwarded-for'] = cookie.credentials['x-forwarded-for'];
```

Comment:
Suggest we define a constant for `'x-forwarded-for'` since it is used many times.

Reply:
created a constant